### ISSUE TITLE
🐛 Bug Report: Deleting conflicting entityRef is not resulting in entity being restored in a new Group.

### ISSUE THREAD

**Issue by dlaird-ovo** [backstage/backstage#17083]
### 📜 Description

When I move an entity between owners in the catalog I get an error like this:

```
Detected conflicting entityRef resource:default/main-db already referenced by file:<dir>/catalog/team-red.yaml and now also file:<dir>/catalog/team-orange.yaml type=plugin
```
I used to look up the UID of the first entity (UI or via GUI)
Then I would DELETE via API
However this is no longer working (Backstage 1.11.x / 1.12.x) - the DELETE succeeds but the replacement entity does not appear in the catalog.

### 👍 Expected behavior

I believe `DELETE`ing the 1st entity should result in the second entity appearing in the catalog (effectively changing the owner)

### 👎 Actual Behavior with Screenshots

I was able to get the UID
I then DELETED and tried to get the item again

```
{
  "error": {
    "name": "NotFoundError",
    "message": "No entity with uid <UID>",
    "stack": "NotFoundError: No entity with uid <UID>\n    at <dir>/node_modules/@backstage/plugin-catalog-backend/dist/index.cjs.js:3619:15\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
  },
  "request": {
    "method": "GET",
    "url": "/entities/by-uid/<UID>"
  },
  "response": {
    "statusCode": 404
  }
}
```
But the second entity is not appearing in the catalog
There are no orphans 
```
curl -X GET localhost:7007/api/catalog/entities?filter=metadata.annotations.backstage.io/orphan=true
```

With matching entity names.
I now have 0 entities of the correct name.

### 👟 Reproduction steps

I ran Backstage locally with a local catalog set of files.
Once running I took an Entity from 1 catalog file and moved to a second file.
I was able to observe the `Conflicting entityRef` message.
I then found the UID of the 1st entityRef (i.e the entity before the move)
I then `DELETED` via the API and confirmed it was deleted.
No replacement entity appeared with a new Owner.
No entity appeared at all.

Restarting Backstage with a clean DB results in the Entity being created in the correct Group - so I believed the file is well-formed.
I am not however able to recreate the DB (easily).

### 📃 Provide the context for the Bug.

Currently changing entity owners in the catalog is not working - and if we resolve the conflicting entityref we lost the entity entirely.

### 🖥️ Your Environment

OS:   Linux 5.15.90.1-microsoft-standard-WSL2 - linux/x64
node: v18.14.2
yarn: 1.22.19
cli:  0.22.5 (installed)
backstage:  1.12.1

Dependencies:
  @backstage/app-defaults                          1.2.1
  @backstage/backend-app-api                       0.4.1
  @backstage/backend-common                        0.18.3
  @backstage/backend-dev-utils                     0.1.1
  @backstage/backend-plugin-api                    0.5.0
  @backstage/backend-tasks                         0.5.0
  @backstage/backend-test-utils                    0.1.35
  @backstage/catalog-client                        1.4.0
  @backstage/catalog-model                         1.2.1
  @backstage/cli-common                            0.1.12
  @backstage/cli                                   0.22.5
  @backstage/config-loader                         1.1.9
  @backstage/config                                1.0.7
  @backstage/core-app-api                          1.6.0
  @backstage/core-components                       0.12.5
  @backstage/core-plugin-api                       1.5.0
  @backstage/errors                                1.1.5
  @backstage/eslint-plugin                         0.1.2
  @backstage/integration-aws-node                  0.1.2
  @backstage/integration-react                     1.1.11
  @backstage/integration                           1.4.3
  @backstage/plugin-api-docs                       0.9.1
  @backstage/plugin-app-backend                    0.3.43
  @backstage/plugin-auth-backend                   0.18.1
  @backstage/plugin-auth-node                      0.2.12
  @backstage/plugin-catalog-backend-module-github  0.2.6
  @backstage/plugin-catalog-backend                1.8.0
  @backstage/plugin-catalog-common                 1.0.12
  @backstage/plugin-catalog-graph                  0.2.28
  @backstage/plugin-catalog-import                 0.9.6
  @backstage/plugin-catalog-node                   1.3.4
  @backstage/plugin-catalog-react                  1.4.0
  @backstage/plugin-catalog                        1.9.0
  @backstage/plugin-circleci                       0.3.16
  @backstage/plugin-events-node                    0.2.4
  @backstage/plugin-explore-common                 0.0.1
  @backstage/plugin-explore-react                  0.0.27
  @backstage/plugin-explore                        0.4.1
  @backstage/plugin-github-actions                 0.5.16
  @backstage/plugin-github-issues                  0.2.5
  @backstage/plugin-github-pull-requests-board     0.1.10
  @backstage/plugin-home                           0.4.32
  @backstage/plugin-kubernetes-backend             0.9.4
  @backstage/plugin-kubernetes-common              0.6.1
  @backstage/plugin-kubernetes                     0.7.9
  @backstage/plugin-org                            0.6.6
  @backstage/plugin-permission-backend             0.5.18
  @backstage/plugin-permission-common              0.7.4
  @backstage/plugin-permission-node                0.7.6
  @backstage/plugin-permission-react               0.4.11
  @backstage/plugin-proxy-backend                  0.2.37
  @backstage/plugin-scaffolder-backend             1.12.0
  @backstage/plugin-scaffolder-common              1.2.6
  @backstage/plugin-scaffolder-node                0.1.1
  @backstage/plugin-scaffolder-react               1.2.0
  @backstage/plugin-scaffolder                     1.12.0
  @backstage/plugin-search-backend-node            1.1.4
  @backstage/plugin-search-backend                 1.2.4
  @backstage/plugin-search-common                  1.2.2
  @backstage/plugin-search-react                   1.5.1
  @backstage/plugin-search                         1.1.1
  @backstage/plugin-tech-radar                     0.6.2
  @backstage/plugin-techdocs-backend               1.6.0
  @backstage/plugin-techdocs-module-addons-contrib 1.0.11
  @backstage/plugin-techdocs-node                  1.6.0
  @backstage/plugin-techdocs-react                 1.1.4
  @backstage/plugin-techdocs                       1.6.0
  @backstage/plugin-user-settings                  0.7.1
  @backstage/release-manifests                     0.0.9
  @backstage/test-utils                            1.2.6
  @backstage/theme                                 0.2.18
  @backstage/types                                 1.0.2
  @backstage/version-bridge                        1.0.3
Done in 0.89s.

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by benjdlambert** [comment]
ping @backstage/catalog-core

**Comment by roman-shandurenko** [comment]
up on this one

**Comment by benjdlambert** [comment]
Ping @backstage/catalog-maintainers

**Comment by freben** [comment]
Alright, thanks for the reproduction case that deals with local files - that does make it easier. I'll try to get some time to try to trigger that happening.

**Comment by Rugvip** [comment]
Wanted to make sure you get a ping about the above comment @backstage/catalog-maintainers in case you want to give it a read

**Comment by camilaibs** [comment]
Ping @backstage/catalog-maintainers

**Comment by dotboris** [comment]
I am experiencing the same issue here. In my case, I used the `GithubEntityProvider` to look for a `catalog-info.yaml` file in every repo. In my setup I support multiple orgs and we're in the process of moving a bunch of repos from one org to another. In some cases this triggers this issue. I think that it might be race-condition-y.

If the old entity (from before the move) exists at the same time as the new entity (from after the move) tries to get ingested then this issue happens and I have to manually fix things to get things back in order.

I managed to mitigate this issue a bit by setting `catalog: { orphanStrategy: delete }` in my config. With that setting things work out most of the time. The general flow (as far as I understand) goes a like this:
1. We have a location that points to `oldorg/repo` and an entity that emitted from that in the catalog
1. We move the repo to `neworg/repo`
1. The old location gets removed & stops emitting the entity
1. Backstage deletes the orphan
2. The entity disappears from the catalog
3. `GithubEntityProvider` finds the repo in the new org
4. It creates a location for it and that location emits the entity
5. The entity shows up in the catalog

I managed to find a simple workaround to fix this issue when it happens. I simply delete the location from the catalog using the backstage UI. The location itself is hard to find in the UI because I can't filter by location targets (https://github.com/backstage/backstage/issues/23098) so I have to look it up through the API and then navigate to it.

What's annoying is that I don't have a good way to figure out which entities are affected by this. The simplest solution I found to this is to use the API to list all locations in the new org and then use the API to delete those.
